src/libraries/System.Private.CoreLib/src/System/IO/AsyncReadLinesEnumerable.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO
+{
+    internal sealed class AsyncReadLinesEnumerable : IAsyncEnumerable<string>
+    {
+        private bool _alreadyGetEnumerator;
+        private readonly int _initialThreadId = Environment.CurrentManagedThreadId;
+        private readonly string _path;
+        private readonly Encoding _encoding;
+        private readonly StreamReader _sr;
+        private readonly CancellationToken _cancellationToken;
+
+        public AsyncReadLinesEnumerable(string path, Encoding encoding, StreamReader sr, CancellationToken cancellationToken)
+        {
+            _path = path;
+            _encoding = encoding;
+            _sr = sr;
+            _cancellationToken = cancellationToken;
+        }
+
+        public IAsyncEnumerator<string> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            StreamReader sr;
+            if (_alreadyGetEnumerator || _initialThreadId != Environment.CurrentManagedThreadId)
+            {
+                sr = new StreamReader(_path, _encoding);
+            }
+            else
+            {
+                sr = _sr;
+                _alreadyGetEnumerator = true;
+            }
+
+            return new AsyncReadLinesEnumerator(sr, CancellationTokenSource.CreateLinkedTokenSource(_cancellationToken, cancellationToken));
+        }
+    }
+
+    internal sealed class AsyncReadLinesEnumerator : IAsyncEnumerator<string>
```

Comment:
We shouldn't implement this manually; as written the implementation is missing out on some important optimizations.  We should instead use the async iterator support provided by the compiler.

Reply:
Fixed, and this iterator has following behaviors, please check if it is our expected ones (they should be same [behaviors](https://github.com/dotnet/runtime/blob/d110d7dd976b36787af31f82644031a20cf50094/src/libraries/System.Private.CoreLib/src/System/IO/ReadLinesIterator.cs#L15) of `ReadLines` though):

Check and throw file related exception if occur upfront before `GetEnumerator`
So that leaking first stream reader if not even `foreach`
Share one stream reader for multiple `enumerator` until first round iteration finishes.